### PR TITLE
Optimized Lagrange Interpolation

### PR DESCRIPTION
**Pull request by sidhantnagpal**
#### Brief description of what is fixed or changed
1. `specialpolys.py: interpolating_poly`: The current implementation (`master`) of `interpolating_poly` uses multiplication for numerator `(n - 1)` times per iteration. In this PR, the multiplication for numerator (`n` times) is precalculated and `1` division takes place per iteration.
2. `polyfuncs.py: interpolate`: The current implementation (`master`) of `interpolate` uses an `O(n**2)` method for interpolation in the special case when `X = list(range(1, n + 1))`. In this PR, this is implemented using an `O(n)` method, utilizing the fact that denominator changes by `1` per step.

Benchmarks (for point 1 mainly):
```
In [1]: %paste
from sympy import *
from sympy.abc import x, y, z

def f(n, x, X='x', Y='y'):
    """Construct Lagrange interpolating polynomial for ``n`` data points. """
    if isinstance(X, str):
        X = symbols("%s:%s" % (X, n))

    if isinstance(Y, str):
        Y = symbols("%s:%s" % (Y, n))

    coeffs = []
    numert = Mul(*[ (x - u) for u in X ])

    for i in range(0, n):
        numer = numert/(x - X[i])
        denom = Mul(*[ (X[i] - X[j]) for j in range(0, n) if i != j ])
        coeffs.append(numer/denom)

    return Add(*[ coeff*y for coeff, y in zip(coeffs, Y) ])

def g(n, x, X='x', Y='y'):
    """Construct Lagrange interpolating polynomial for ``n`` data points. """
    if isinstance(X, str):
        X = symbols("%s:%s" % (X, n))

    if isinstance(Y, str):
        Y = symbols("%s:%s" % (Y, n))

    coeffs = []

    for i in range(0, n):
        numer = []
        denom = []

        for j in range(0, n):
            if i == j:
                continue

            numer.append(x - X[j])
            denom.append(X[i] - X[j])

        numer = Mul(*numer)
        denom = Mul(*denom)

        coeffs.append(numer/denom)

    return Add(*[ coeff*y for coeff, y in zip(coeffs, Y) ])

x0, x1, x2, x3, x4, x5, y0, y1, y2, y3, y4, y5 = symbols('x:6, y:6')

## -- End pasted text --

In [2]: %timeit f(5, x)
1000 loops, best of 3: 1.49 ms per loop

In [3]: %timeit g(5, x)
1000 loops, best of 3: 1.7 ms per loop

In [4]: %timeit f(6, x)
100 loops, best of 3: 1.96 ms per loop

In [5]: %timeit g(6, x)
100 loops, best of 3: 2.37 ms per loop
```

### CODE REVIEW

File: sympy/polys/polyfuncs.py [hunk 1]
```diff
@@ -239,17 +239,26 @@ def interpolate(data, x):
...
         if isinstance(data[0], tuple):
             X, Y = list(zip(*data))
+            poly = interpolating_poly(n, x, X, Y)
         else:
-            X = list(range(1, n + 1))
             Y = list(data)
 
-    poly = interpolating_poly(n, x, X, Y)
+            numert = Mul(*[ (x - i) for i in range(1, n + 1) ])
+            denom = Mul(*[ (1 - i) for i in range(2, n + 1) ])
```
Comment (jksuom) [inline sympy/polys/polyfuncs.py]: This looks like a factorial. Could that be used?

BTW, PEP 8 is not always strictly followed in SymPy, but this rule is mostly applied:

    Avoid extraneous whitespace in the following situations:

    Immediately inside parentheses, brackets or braces.

    Yes: spam(ham[1], {eggs: 2})
    No:  spam( ham[ 1 ], { eggs: 2 } )

Comment (sidhantnagpal) [reply sympy/polys/polyfuncs.py]: > This looks like a factorial. Could that be used?

Yes, with care for negative sign. Will do.

> BTW, PEP 8 is not always strictly followed in SymPy, but this rule is mostly applied.

I remember seeing this type of formatting used in SymPy. 
Should I remove the spaces before `]` and after `[`?

Comment (jksuom) [reply sympy/polys/polyfuncs.py]: I have also seen that type of formatting but that should probably be avoided in new code. (So please remove the spaces.)

Comment (sidhantnagpal) [reply sympy/polys/polyfuncs.py]: > I remember seeing this type of formatting used in SymPy.

In line `138` of `polyfuncs.py`: [here](https://github.com/sidhantnagpal/sympy/blob/lagrange-interpolate/sympy/polys/polyfuncs.py#L138)

Comment (sidhantnagpal) [reply sympy/polys/polyfuncs.py]: > I have also seen that type of formatting but that should probably be avoided in new code. (So please remove the spaces.)

Oh, okay. So should I remove the spaces throughout the file (for existing lines as well)?

**Review by sidhantnagpal — COMMENTED** [review]
Made the required changes.

**Comment by jksuom** [comment]
Thanks, this looks good.

**Comment by sidhantnagpal** [comment]
Thank you @jksuom.

**Pull request merged**